### ISSUE TITLE
Project file format

### ISSUE THREAD

**Issue by jelovirt** [dita-ot/dita-ot#3144]
## Description

Add support for input source project file, that defines

- input files
- filters
- transtypes
- parameters

The project file doesn't add new processing features, only allows defining all arguments in a single file.

**Comment by jelovirt** [comment]
Possible ways to use the project file:

  Generate all publications for each deliverable

  ```bash
  $ dita --project=docs.project
  ```

  You can specify project and only some deliverable and some publication

  ```bash
  $ dita --project=docs.project --deliverable=inline --publication=html
  ```

  or maybe shorthand

  ```bash
  $ dita --project=docs.project --publication=inline/html
  ```

**Comment by drmacro** [comment]
One requirement for cross-deliverable linking is having a single object or artifact that represents the deliverable as a tuple of root map, transformation specification details, and applied filtering conditions, that is root map, transtype, and ditavals, where "transtype" encapsulates all the transformation-specific configuration details.

In the context of cross-deliverable link generation it is necessary to define a set of related deliverables that are expected to have links among them such that you can process them as a coordinated set, whatever "coordinated means" (e.g., sharing set of "public" element-to-deliverable anchor mappings that enable generation of working cross-deliverable links).

I'm not sure if these requirements need to be reflected in the project file design--it would be possible, for example, to define a map specialization that serves as the deliverable definition and another that serves as the "related set of deliverables" definition, either of which could then be referenced from a project file as you've described them.

But I wanted to make sure that these requirements were at least known as you're thinking about project and the project file design.

**Comment by raducoravu** [comment]
So for each input file you have a set of filters. For each (inputFile, filter) pair you have a set of possible transtypes. And for each (inputFile, filter, transtype) pair you have a set of defined transformation parameters (or a reference to a properties file defining them).
I would prefer an XML format for this, @jelovirt do you incline towards using JSON?

**Comment by xephon2** [comment]
I really like the approach of @lionelmoi to store all the publication data in `<data>` elements on the publication map, he has presented at the DITA-OT-Day in Rotterdam. This is very user friendly, as a "normal" technical writer should only edit DITA maps and topics and no other configuration or build files. 
Maybe you can consider implementing a processing that automatically reads `<data>` elements like `<data name="custom.property">value</data>` and store the values in properties, if you rework the "project" approach.

**Comment by lionelmoi** [comment]
We have a specific approach to defining multiple maps to be published at the same time.
We use `@scope="peer"` this way :
```
   <topichead navtitle="peers" toc="no" processing-role="resource-only">
      <mapref href="release_notes_ADVANCED_publi.ditamap" scope="peer">
         <topicmeta>
            <navtitle>Release notes</navtitle>
         </topicmeta>
      </mapref>
      <mapref href="_tutorials_publi.ditamap" scope="peer">
         <topicmeta>
            <navtitle>Tutorials</navtitle>
         </topicmeta>
      </mapref>
      <mapref href="recommendations_publi.ditamap" scope="peer">
         <topicmeta>
            <navtitle>Recommendations</navtitle>
         </topicmeta>
      </mapref>
   </topichead>
```

This example is from the User guide ditamap.

If the `peers` parameter is set to `yes` in my build, it reads the source ditamap and picks up each referenced file to add it to the list of input maps.

The transtype of each input files is picked up from the ditamap itself.

Cross-links are resolved based on the `<navtitle>` of the map which corresponds to the actual title and defines the output folder. This is not the most developed aspect of my system, we usually resort to using `@scope="external"` to link from one publication output to another.

This could obviously be done through a `<data>` (might be better) or through a `<mapref>` specialization.

**Comment by robander** [comment]
@raducoravu about this: 

> So for each input file you have a set of filters. For each (inputFile, filter) pair you have a set of possible transtypes.

I think that's often the case - but not the only common case. For us it is very common to have filter conditions that are related to the transform type -- for example, online includes almost everything, PDF filters out a lot, online includes peer links that may be handled differently in PDF, etc.

So for us the most common scenario now is slightly different: For each input file we have one or more set of common filter conditions. For each (inputFile, filter) set we also have a set of possible transform types and additional filter conditions.

Small difference but I wouldn't want the project file to assume any map always uses the same conditions.

**Comment by jelovirt** [comment]
Implemented in #3243
